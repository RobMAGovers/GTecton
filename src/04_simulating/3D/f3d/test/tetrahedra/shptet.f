	subroutine SHPTET (X,Y,Z,XL,DET,SH,IERR,IFLAG)
	USE CONSTANTS, only: sixth, one
*
c Computes shape functions and their derivatives for linear tetrahedra (simplex
c element).
*
	implicit none
	integer NEN,NSD
	parameter (NEN=4,NSD=3)
c-pass
	integer IERR,IFLAG
	double precision X,Y,Z,XL,SH,DET
	dimension XL(NSD,NEN),SH(4,NEN)
c-comm
c-locl
	double precision A1,A2,A3,A4,B1,B2,B3,B4,C1,C2,C3,C4,D1,D2,D3,
  >	 D4,RDET
c-init
	IERR = 0
*
	A1 = XL(1,2)*(XL(2,3)*XL(3,4)-XL(3,3)*XL(2,4))-
  >	     XL(1,3)*(XL(2,2)*XL(3,4)-XL(3,2)*XL(2,4))+
  >	     XL(1,4)*(XL(2,2)*XL(3,3)-XL(3,2)*XL(2,3))
	A2 = XL(1,1)*(XL(3,3)*XL(2,4)-XL(2,3)*XL(3,4))-
  >	     XL(1,3)*(XL(3,1)*XL(2,4)-XL(2,1)*XL(3,4))+
  >	     XL(1,4)*(XL(3,1)*XL(2,3)-XL(2,1)*XL(3,3))
	A3 = XL(1,1)*(XL(2,2)*XL(3,4)-XL(3,2)*XL(2,4))-
  >	     XL(1,2)*(XL(2,1)*XL(3,4)-XL(3,1)*XL(2,4))+
  >	     XL(1,4)*(XL(2,1)*XL(3,2)-XL(3,1)*XL(2,2))
	A4 = XL(1,1)*(XL(3,2)*XL(2,3)-XL(2,2)*XL(3,3))-
  >	     XL(1,2)*(XL(3,1)*XL(2,3)-XL(2,1)*XL(3,3))+
  >	     XL(1,3)*(XL(3,1)*XL(2,2)-XL(2,1)*XL(3,2))
	B1 = XL(3,3)*XL(2,4)-XL(2,3)*XL(3,4) -
  >	     XL(3,2)*XL(2,4)+XL(2,2)*XL(3,4) +
  >	     XL(3,2)*XL(2,3)-XL(2,2)*XL(3,3)
	B2 = XL(2,3)*XL(3,4)-XL(3,3)*XL(2,4) -
  >	     XL(2,1)*XL(3,4)+XL(3,1)*XL(2,4) +
  >	     XL(2,1)*XL(3,3)-XL(3,1)*XL(2,3)
	B3 = XL(3,2)*XL(2,4)-XL(2,2)*XL(3,4) -
  >	     XL(3,1)*XL(2,4)+XL(2,1)*XL(3,4) +
  >	     XL(3,1)*XL(2,2)-XL(2,1)*XL(3,2)
	B4 = XL(2,2)*XL(3,3)-XL(3,2)*XL(2,3) -
  >	     XL(2,1)*XL(3,3)+XL(3,1)*XL(2,3) +
  >	     XL(2,1)*XL(3,2)-XL(3,1)*XL(2,2)
	C1 = XL(1,3)*XL(3,4)-XL(3,3)*XL(1,4) -
  >	     XL(1,2)*XL(3,4)+XL(3,2)*XL(1,4) +
  >	     XL(1,2)*XL(3,3)-XL(3,2)*XL(1,3)
	C2 = XL(3,3)*XL(1,4)-XL(1,3)*XL(3,4) -
  >	     XL(3,1)*XL(1,4)+XL(1,1)*XL(3,4) +
  >	     XL(3,1)*XL(1,3)-XL(1,1)*XL(3,3)
	C3 = XL(1,2)*XL(3,4)-XL(3,2)*XL(1,4) -
  >	     XL(1,1)*XL(3,4)+XL(3,1)*XL(1,4) +
  >	     XL(1,1)*XL(3,2)-XL(3,1)*XL(1,2)
	C4 = XL(3,2)*XL(1,3)-XL(1,2)*XL(3,3) -
  >	     XL(3,1)*XL(1,3)+XL(1,1)*XL(3,3) +
  >	     XL(3,1)*XL(1,2)-XL(1,1)*XL(3,2)
	D1 = XL(2,3)*XL(1,4)-XL(1,3)*XL(2,4) -
  >	     XL(2,2)*XL(1,4)+XL(1,2)*XL(2,4) +
  >	     XL(2,2)*XL(1,3)-XL(1,2)*XL(2,3)
	D2 = XL(1,3)*XL(2,4)-XL(2,3)*XL(1,4) -
  >	     XL(1,1)*XL(2,4)+XL(2,1)*XL(1,4) +
  >	     XL(1,1)*XL(2,3)-XL(2,1)*XL(1,3)
	D3 = XL(2,2)*XL(1,4)-XL(1,2)*XL(2,4) -
  >	     XL(2,1)*XL(1,4)+XL(1,1)*XL(2,4) +
  >	     XL(2,1)*XL(1,2)-XL(1,1)*XL(2,2)
	D4 = XL(1,1)*XL(2,2)-XL(2,1)*XL(1,2) -
  >	     XL(1,1)*XL(2,3)+XL(2,1)*XL(1,3) +
  >	     XL(1,2)*XL(2,3)-XL(2,2)*XL(1,3)
*
	DET = A1+A2+A3+A4
	if (DET.le.0d0) goto 1000
	RDET = ONE/DET
*
	if (IFLAG.eq.1) then
c	    Compute x-, y-, and z-derivatives of shape functions
	    SH(1,1) = B1*RDET
	    SH(1,2) = B2*RDET
	    SH(1,3) = B3*RDET
	    SH(1,4) = B4*RDET
	    SH(2,1) = C1*RDET
	    SH(2,2) = C2*RDET
	    SH(2,3) = C3*RDET
	    SH(2,4) = C4*RDET
	    SH(3,1) = D1*RDET
	    SH(3,2) = D2*RDET
	    SH(3,3) = D3*RDET
	    SH(3,4) = D4*RDET
	else
	    SH(4,1) = (A1+B1*X+C1*Y+D1*Z)*RDET
	    SH(4,2) = (A2+B2*X+C2*Y+D2*Z)*RDET
	    SH(4,3) = (A3+B3*X+C3*Y+D3*Z)*RDET
	    SH(4,4) = (A4+B4*X+C4*Y+D4*Z)*RDET
	endif
*
c	compute volume
	DET = DET*SIXTH
*
	return
*
1000	write(0,1) DET
 1	format(///1x,'Shape function fails! Determinant is ',1PE20.4)
	IERR = 1
	return
	end
