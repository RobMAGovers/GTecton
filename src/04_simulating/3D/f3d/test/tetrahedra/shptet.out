 Input coordinates
 XL(1)=   -0.500000   -0.500000   -0.500000
 XL(2)=    0.500000   -0.500000   -0.500000
 XL(3)=   -0.500000    0.500000   -0.500000
 XL(4)=   -0.500000   -0.500000    0.500000

 Rotate XL over  30. degrees along Z-axis
 XL(1)=   -0.183013   -0.683013   -0.500000
 XL(2)=    0.683013   -0.183013   -0.500000
 XL(3)=   -0.683013    0.183013   -0.500000
 XL(4)=   -0.183013   -0.683013    0.500000

 DET=  0.166666667
 temperature gradient:
1.665335E-16   1.110223E-16    1.00000
 Volume integral of T =   0.208333333
 nodal shape function values:
 1.00000       3.295975E-17  -4.287511E-17    0.00000
 0.00000        1.00000       3.296483E-17    0.00000
 0.00000      -4.288697E-17    1.00000        0.00000
  -1.038124E-17   3.295975E-17  -4.288528E-17    1.00000
 element center and shape function values:
  -9.150635E-02  -0.341506      -0.250000
0.250000       0.250000       0.250000       0.250000
